File: tests/dotnet/StepDefinitions/IndexerSteps.cs
/**
 * Indexer GraphQL Client Step Definitions
 *
 * Implements behavioral tests for Aptos Indexer GraphQL API client.
 * Note: Network-dependent tests use mock responses for offline testing.
 */
using Reqnroll;
using FluentAssertions;
using Aptos;
using Aptos.Specs.Support;

namespace Aptos.Specs.StepDefinitions;

[Binding]
public class IndexerSteps
{
    private readonly TestWorld _world;

    public IndexerSteps(TestWorld world)
    {
        _world = world;
    }

    // =========================================================================
    // Client Configuration
    // =========================================================================

    [When("I create an indexer client for mainnet")]
    public void WhenICreateAnIndexerClientForMainnet()
    {
        _world.TestVectors["indexerNetwork"] = "mainnet";
        _world.TestVectors["expectedIndexerUrl"] = "https://indexer.mainnet.aptoslabs.com/v1/graphql";
    }

    [When("I create an indexer client for testnet")]
    public void WhenICreateAnIndexerClientForTestnet()
    {
        _world.TestVectors["indexerNetwork"] = "testnet";
        _world.TestVectors["expectedIndexerUrl"] = "https://indexer.testnet.aptoslabs.com/v1/graphql";
    }

    [Given("a custom indexer URL")]
    public void GivenACustomIndexerURL()
    {
        _world.TestVectors["customIndexerUrl"] = "https://custom-indexer.example.com/v1/graphql";
    }

    [When("I create an indexer client with the custom URL")]
    public void WhenICreateAnIndexerClientWithTheCustomURL()
    {
        var customUrl = _world.TestVectors["customIndexerUrl"] as string;
        _world.TestVectors["configuredIndexerUrl"] = customUrl;
    }

    [Then("the indexer client should use that URL")]
    public void ThenTheIndexerClientShouldUseThatURL()
    {
        var customUrl = _world.TestVectors["customIndexerUrl"] as string;
        var configuredUrl = _world.TestVectors["configuredIndexerUrl"] as string;
        configuredUrl.Should().Be(customUrl);
    }

    [Given("an API key for the indexer")]
    public void GivenAnAPIKeyForTheIndexer()
    {
        _world.TestVectors["indexerApiKey"] = "test-api-key-12345";
    }

    [When("I create an indexer client with the key")]
    public void WhenICreateAnIndexerClientWithTheKey()
    {
        var apiKey = _world.TestVectors["indexerApiKey"] as string;
        _world.TestVectors["configuredApiKey"] = apiKey;
    }

    [Then("requests should include the API key header")]
    public void ThenRequestsShouldIncludeTheAPIKeyHeader()
    {
        _world.TestVectors["configuredApiKey"].Should().NotBeNull();
    }

    // =========================================================================
    // Raw GraphQL Queries
    // =========================================================================

    [Given("an indexer client")]
    public void GivenAnIndexerClient()
    {
        _world.TestVectors["indexerAvailable"] = true;
    }

    [Given("a GraphQL query string")]
    public void GivenAGraphQLQueryString()
    {
        _world.TestVectors["graphqlQuery"] = @"
            query GetLedgerInfo {
                ledger_infos {
                    chain_id
                }
            }
        ";
    }

    [When("I execute the query")]
    public void WhenIExecuteTheQuery()
    {
        // Mock execution
        _world.TestVectors["queryResult"] = new Dictionary<string, object>
        {
            { "data", new Dictionary<string, object>
                {
                    { "ledger_infos", new List<object> { new Dictionary<string, object> { { "chain_id", 2 } } } }
                }
            }
        };
        _world.Result = _world.TestVectors["queryResult"];
    }

    [Then("I should receive the query result")]
    public void ThenIShouldReceiveTheQueryResult()
    {
        var result = _world.TestVectors["queryResult"] as Dictionary<string, object>;
        result.Should().NotBeNull();
        result.Should().ContainKey("data");
    }

    [Given("a GraphQL query with variables")]
    public void GivenAGraphQLQueryWithVariables()
    {
        _world.TestVectors["graphqlQuery"] = @"
            query GetAccountTokens($address: String!) {
                current_token_ownerships_v2(where: {owner_address: {_eq: $address}}) {
                    token_data_id
                    amount
                }
            }
        ";
    }

    [Given("variable values")]
    public void GivenVariableValues()
    {
        _world.TestVectors["queryVariables"] = new Dictionary<string, object> { { "address", "0x1" } };
    }

    [When("I execute the query with variables")]
    public void WhenIExecuteTheQueryWithVariables()
    {
        _world.TestVectors["queryResult"] = new Dictionary<string, object>
        {
            { "data", new Dictionary<string, object>
                {
                    { "current_token_ownerships_v2", new List<object>
                        {
                            new Dictionary<string, object> { { "token_data_id", "0x123" }, { "amount", 1 } }
                        }
                    }
                }
            }
        };
        _world.Result = _world.TestVectors["queryResult"];
    }

    [Then("the variables should be substituted")]
    public void ThenTheVariablesShouldBeSubstituted()
    {
        var result = _world.TestVectors["queryResult"] as Dictionary<string, object>;
        var data = result!["data"] as Dictionary<string, object>;
        var tokens = data!["current_token_ownerships_v2"] as List<object>;
        tokens.Should().NotBeNull();
    }

    [Given("an invalid GraphQL query")]
    public void GivenAnInvalidGraphQLQuery()
    {
        _world.TestVectors["graphqlQuery"] = "{ invalid query syntax";
    }

    [When("I execute the GraphQL query")]
    public void WhenIExecuteTheGraphQLQuery()
    {
        _world.SetError(new Exception("GraphQL syntax error"));
        _world.TestVectors["graphqlError"] = new Dictionary<string, object>
        {
            { "errors", new List<object> { new Dictionary<string, object> { { "message", "Syntax error" } } } }
        };
    }

    [Then("I should receive a GraphQL error")]
    public void ThenIShouldReceiveAGraphQLError()
    {
        _world.Error.Should().NotBeNull();
    }

    [Then("the error should contain the error message")]
    public void ThenTheErrorShouldContainTheErrorMessage()
    {
        _world.Error!.Message.Should().NotBeEmpty();
    }

    // =========================================================================
    // Account Tokens (NFTs)
    // =========================================================================

    [Given("an account address with NFTs")]
    public void GivenAnAccountAddressWithNFTs()
    {
        _world.TestVectors["accountAddress"] = AccountAddress.FromString("0x1");
    }

    [When("I query current tokens for the account")]
    public void WhenIQueryCurrentTokensForTheAccount()
    {
        _world.TestVectors["tokensResult"] = new List<Dictionary<string, object>>
        {
            new()
            {
                { "token_data_id", "0x123::collection::token1" },
                { "collection_name", "Test Collection" },
                { "token_name", "Token 1" },
                { "amount", 1 }
            }
        };
        _world.Result = _world.TestVectors["tokensResult"];
    }

    [Then("I should receive a list of tokens")]
    public void ThenIShouldReceiveAListOfTokens()
    {
        var tokens = _world.TestVectors["tokensResult"] as List<Dictionary<string, object>>;
        tokens.Should().NotBeNull();
    }

    [Then("each token should have collection info")]
    public void ThenEachTokenShouldHaveCollectionInfo()
    {
        var tokens = _world.TestVectors["tokensResult"] as List<Dictionary<string, object>>;
        foreach (var token in tokens!)
        {
            token.Should().ContainKey("collection_name");
        }
    }

    [Then("each token should have token_data_id")]
    public void ThenEachTokenShouldHaveTokenDataId()
    {
        var tokens = _world.TestVectors["tokensResult"] as List<Dictionary<string, object>>;
        foreach (var token in tokens!)
        {
            token.Should().ContainKey("token_data_id");
        }
    }

    [Given("an account with many NFTs")]
    public void GivenAnAccountWithManyNFTs()
    {
        _world.TestVectors["accountAddress"] = AccountAddress.FromString("0x1");
        _world.TestVectors["totalNfts"] = 25;
    }

    [When("I query tokens with limit {int} and offset {int}")]
    public void WhenIQueryTokensWithLimitAndOffset(int limit, int offset)
    {
        var totalNfts = (int)_world.TestVectors["totalNfts"];
        var tokens = new List<Dictionary<string, object>>();
        for (int i = offset; i < Math.Min(offset + limit, totalNfts); i++)
        {
            tokens.Add(new Dictionary<string, object> { { "token_data_id", $"token_{i}" }, { "amount", 1 } });
        }
        _world.TestVectors["tokensResult"] = tokens;
        _world.TestVectors["queryLimit"] = limit;
    }

    [Then("I should receive at most {int} tokens")]
    public void ThenIShouldReceiveAtMostTokens(int limit)
    {
        var tokens = _world.TestVectors["tokensResult"] as List<Dictionary<string, object>>;
        tokens!.Count.Should().BeLessThanOrEqualTo(limit);
    }

    [When("I query with offset {int}")]
    public void WhenIQueryWithOffset(int offset)
    {
        var totalNfts = _world.TestVectors.TryGetValue("totalNfts", out var tn) ? (int)tn : 25;
        var limit = _world.TestVectors.TryGetValue("queryLimit", out var ql) ? (int)ql : 10;
        var tokens = new List<Dictionary<string, object>>();
        for (int i = offset; i < Math.Min(offset + limit, totalNfts); i++)
        {
            tokens.Add(new Dictionary<string, object> { { "token_data_id", $"token_{i}" }, { "amount", 1 } });
        }
        _world.TestVectors["tokensResult"] = tokens;
    }

    [Then("I should receive the next page")]
    public void ThenIShouldReceiveTheNextPage()
    {
        var tokens = _world.TestVectors["tokensResult"] as List<Dictionary<string, object>>;
        tokens!.Count.Should().BeGreaterThan(0);
    }

    [Given("an account with an NFT")]
    public void GivenAnAccountWithAnNFT()
    {
        _world.TestVectors["accountAddress"] = AccountAddress.FromString("0x1");
    }

    [When("I query the token")]
    public void WhenIQueryTheToken()
    {
        _world.TestVectors["tokenResult"] = new Dictionary<string, object>
        {
            { "token_name", "My NFT" },
            { "collection_name", "My Collection" },
            { "token_uri", "https://example.com/token/1" },
            { "amount", 1 }
        };
        _world.Result = _world.TestVectors["tokenResult"];
    }

    [Then("I should see token_name")]
    public void ThenIShouldSeeTokenName()
    {
        var token = _world.TestVectors["tokenResult"] as Dictionary<string, object>;
        token.Should().ContainKey("token_name");
    }

    [Then("I should see collection_name")]
    public void ThenIShouldSeeCollectionName()
    {
        var token = _world.TestVectors["tokenResult"] as Dictionary<string, object>;
        token.Should().ContainKey("collection_name");
    }

    [Then("I should see token_uri")]
    public void ThenIShouldSeeTokenUri()
    {
        var token = _world.TestVectors["tokenResult"] as Dictionary<string, object>;
        token.Should().ContainKey("token_uri");
    }

    [Then("I should see amount")]
    public void ThenIShouldSeeAmount()
    {
        var token = _world.TestVectors["tokenResult"] as Dictionary<string, object>;
        token.Should().ContainKey("amount");
    }

    [Given("an account with no NFTs")]
    public void GivenAnAccountWithNoNFTs()
    {
        _world.TestVectors["accountAddress"] = AccountAddress.FromString("0x999");
    }

    [When("I query current tokens")]
    public void WhenIQueryCurrentTokens()
    {
        _world.TestVectors["tokensResult"] = new List<Dictionary<string, object>>();
    }

    [Then("I should receive an empty list")]
    public void ThenIShouldReceiveAnEmptyList()
    {
        var tokens = _world.TestVectors["tokensResult"] as List<Dictionary<string, object>>;
        tokens.Should().NotBeNull();
        tokens!.Count.Should().Be(0);
    }

    // =========================================================================
    // Fungible Asset Balances
    // =========================================================================

    [When("I query fungible asset balances")]
    public void WhenIQueryFungibleAssetBalances()
    {
        _world.TestVectors["fungibleAssetBalances"] = new List<Dictionary<string, object>>
        {
            new() { { "asset_type", "0x1::aptos_coin::AptosCoin" }, { "amount", "1000000000" } }
        };
        _world.Result = _world.TestVectors["fungibleAssetBalances"];
    }

    [Then("I should receive a list of fungible asset balances")]
    public void ThenIShouldReceiveAListOfFungibleAssetBalances()
    {
        var balances = _world.TestVectors["fungibleAssetBalances"] as List<Dictionary<string, object>>;
        balances.Should().NotBeNull();
    }

    [Then("each balance should have asset_type and amount")]
    public void ThenEachBalanceShouldHaveAssetTypeAndAmount()
    {
        var balances = _world.TestVectors["fungibleAssetBalances"] as List<Dictionary<string, object>>;
        foreach (var balance in balances!)
        {
            balance.Should().ContainKey("asset_type");
            balance.Should().ContainKey("amount");
        }
    }

    [Given("a specific fungible asset address")]
    public void GivenASpecificFungibleAssetAddress()
    {
        _world.TestVectors["fungibleAssetAddress"] = "0x1::aptos_coin::AptosCoin";
    }

    [When("I query that specific balance")]
    public void WhenIQueryThatSpecificBalance()
    {
        _world.TestVectors["specificBalance"] = new Dictionary<string, object>
        {
            { "asset_type", "0x1::aptos_coin::AptosCoin" },
            { "amount", "1000000000" }
        };
        _world.Result = _world.TestVectors["specificBalance"];
    }

    [Then("I should receive only that asset's balance")]
    public void ThenIShouldReceiveOnlyThatAssetsBalance()
    {
        var balance = _world.TestVectors["specificBalance"] as Dictionary<string, object>;
        balance.Should().ContainKey("asset_type");
    }

    [Given("a fungible asset address")]
    public void GivenAFungibleAssetAddress()
    {
        _world.TestVectors["fungibleAssetAddress"] = "0x1::aptos_coin::AptosCoin";
    }

    [When("I query the asset metadata")]
    public void WhenIQueryTheAssetMetadata()
    {
        _world.TestVectors["assetMetadata"] = new Dictionary<string, object>
        {
            { "name", "Aptos Coin" },
            { "symbol", "APT" },
            { "decimals", 8 }
        };
        _world.Result = _world.TestVectors["assetMetadata"];
    }

    [Then("I should see name, symbol, and decimals")]
    public void ThenIShouldSeeNameSymbolAndDecimals()
    {
        var metadata = _world.TestVectors["assetMetadata"] as Dictionary<string, object>;
        metadata.Should().ContainKey("name");
        metadata.Should().ContainKey("symbol");
        metadata.Should().ContainKey("decimals");
    }

    // =========================================================================
    // Transaction History
    // =========================================================================

    [When("I query account transaction history")]
    public void WhenIQueryAccountTransactionHistory()
    {
        _world.TestVectors["transactionHistory"] = new List<Dictionary<string, object>>
        {
            new() { { "version", "1" }, { "type", "user_transaction" }, { "success", true } },
            new() { { "version", "2" }, { "type", "user_transaction" }, { "success", true } }
        };
        _world.Result = _world.TestVectors["transactionHistory"];
    }

    [Then("I should receive a list of indexed transactions")]
    public void ThenIShouldReceiveAListOfIndexedTransactions()
    {
        var txns = _world.TestVectors["transactionHistory"] as List<Dictionary<string, object>>;
        txns.Should().NotBeNull();
    }

    [Then("each transaction should have version and type")]
    public void ThenEachTransactionShouldHaveVersionAndType()
    {
        var txns = _world.TestVectors["transactionHistory"] as List<Dictionary<string, object>>;
        foreach (var txn in txns!)
        {
            txn.Should().ContainKey("version");
            txn.Should().ContainKey("type");
        }
    }

    [When("I query with limit {int}")]
    public void WhenIQueryWithLimit(int limit)
    {
        var txns = new List<Dictionary<string, object>>();
        for (int i = 0; i < limit; i++)
        {
            txns.Add(new Dictionary<string, object> { { "version", i.ToString() }, { "type", "user_transaction" } });
        }
        _world.TestVectors["transactionHistory"] = txns;
    }

    [Then("I should receive at most {int} indexed transactions")]
    public void ThenIShouldReceiveAtMostIndexedTransactions(int limit)
    {
        var txns = _world.TestVectors["transactionHistory"] as List<Dictionary<string, object>>;
        txns!.Count.Should().BeLessThanOrEqualTo(limit);
    }

    [Given("a user_transaction type filter")]
    public void GivenAUserTransactionTypeFilter()
    {
        _world.TestVectors["transactionTypeFilter"] = "user_transaction";
    }

    [When("I query with the filter")]
    public void WhenIQueryWithTheFilter()
    {
        _world.TestVectors["transactionHistory"] = new List<Dictionary<string, object>>
        {
            new() { { "version", "1" }, { "type", "user_transaction" } }
        };
    }

    [Then("all transactions should be user_transactions")]
    public void ThenAllTransactionsShouldBeUserTransactions()
    {
        var txns = _world.TestVectors["transactionHistory"] as List<Dictionary<string, object>>;
        foreach (var txn in txns!)
        {
            txn["type"].Should().Be("user_transaction");
        }
    }

    // =========================================================================
    // Collections
    // =========================================================================

    [Given("a known collection address")]
    public void GivenAKnownCollectionAddress()
    {
        _world.TestVectors["collectionAddress"] = "0x123::my_collection::MyCollection";
    }

    [When("I query the collection")]
    public void WhenIQueryTheCollection()
    {
        _world.TestVectors["collectionResult"] = new Dictionary<string, object>
        {
            { "collection_name", "My Collection" },
            { "creator_address", "0x123" },
            { "description", "A test collection" },
            { "current_supply", 100 }
        };
        _world.Result = _world.TestVectors["collectionResult"];
    }

    [Then("I should receive collection details")]
    public void ThenIShouldReceiveCollectionDetails()
    {
        var collection = _world.TestVectors["collectionResult"] as Dictionary<string, object>;
        collection.Should().NotBeNull();
    }

    [Then("I should see creator_address")]
    public void ThenIShouldSeeCreatorAddress()
    {
        var collection = _world.TestVectors["collectionResult"] as Dictionary<string, object>;
        collection.Should().ContainKey("creator_address");
    }

    [Then("I should see current_supply")]
    public void ThenIShouldSeeCurrentSupply()
    {
        var collection = _world.TestVectors["collectionResult"] as Dictionary<string, object>;
        collection.Should().ContainKey("current_supply");
    }

    [When("I query tokens in the collection")]
    public void WhenIQueryTokensInTheCollection()
    {
        _world.TestVectors["collectionTokens"] = new List<Dictionary<string, object>>
        {
            new() { { "token_data_id", "token1" }, { "token_name", "Token 1" } },
            new() { { "token_data_id", "token2" }, { "token_name", "Token 2" } }
        };
        _world.Result = _world.TestVectors["collectionTokens"];
    }

    [Then("I should receive tokens belonging to that collection")]
    public void ThenIShouldReceiveTokensBelongingToThatCollection()
    {
        var tokens = _world.TestVectors["collectionTokens"] as List<Dictionary<string, object>>;
        tokens.Should().NotBeNull();
        tokens!.Count.Should().BeGreaterThan(0);
    }

    [When("I query collection metadata")]
    public void WhenIQueryCollectionMetadata()
    {
        _world.TestVectors["collectionMetadata"] = new Dictionary<string, object>
        {
            { "name", "My Collection" },
            { "description", "A test collection" },
            { "uri", "https://example.com/collection" }
        };
        _world.Result = _world.TestVectors["collectionMetadata"];
    }

    [Then("I should see collection name and description")]
    public void ThenIShouldSeeCollectionNameAndDescription()
    {
        var metadata = _world.TestVectors["collectionMetadata"] as Dictionary<string, object>;
        metadata.Should().ContainKey("name");
        metadata.Should().ContainKey("description");
    }

    [Then("I should see collection URI")]
    public void ThenIShouldSeeCollectionURI()
    {
        var metadata = _world.TestVectors["collectionMetadata"] as Dictionary<string, object>;
        metadata.Should().ContainKey("uri");
    }

    // =========================================================================
    // Events
    // =========================================================================

    [Given("an event type {string}")]
    public void GivenAnEventType(string eventType)
    {
        _world.TestVectors["eventType"] = eventType;
    }

    [When("I query events of that type")]
    public void WhenIQueryEventsOfThatType()
    {
        _world.TestVectors["eventsResult"] = new List<Dictionary<string, object>>
        {
            new() { { "type", "0x1::coin::DepositEvent" }, { "data", new Dictionary<string, object> { { "amount", "100" } } } }
        };
        _world.Result = _world.TestVectors["eventsResult"];
    }

    [Then("I should receive matching events")]
    public void ThenIShouldReceiveMatchingEvents()
    {
        var events = _world.TestVectors["eventsResult"] as List<Dictionary<string, object>>;
        events.Should().NotBeNull();
    }

    [When("I query events for the account")]
    public void WhenIQueryEventsForTheAccount()
    {
        WhenIQueryEventsOfThatType();
    }

    [Then("I should receive events involving that account")]
    public void ThenIShouldReceiveEventsInvolvingThatAccount()
    {
        ThenIShouldReceiveMatchingEvents();
    }

    [Then("each event should have type and data")]
    public void ThenEachEventShouldHaveTypeAndData()
    {
        var events = _world.TestVectors["eventsResult"] as List<Dictionary<string, object>>;
        foreach (var evt in events!)
        {
            evt.Should().ContainKey("type");
            evt.Should().ContainKey("data");
        }
    }

    // =========================================================================
    // Legacy Coins
    // =========================================================================

    [When("I query coin balances")]
    public void WhenIQueryCoinBalances()
    {
        _world.TestVectors["coinBalances"] = new List<Dictionary<string, object>>
        {
            new() { { "coin_type", "0x1::aptos_coin::AptosCoin" }, { "amount", "1000000000" } }
        };
        _world.Result = _world.TestVectors["coinBalances"];
    }

    [Then("I should receive legacy coin balances")]
    public void ThenIShouldReceiveLegacyCoinBalances()
    {
        var balances = _world.TestVectors["coinBalances"] as List<Dictionary<string, object>>;
        balances.Should().NotBeNull();
    }

    [When("I query coin activities")]
    public void WhenIQueryCoinActivities()
    {
        _world.TestVectors["coinActivities"] = new List<Dictionary<string, object>>
        {
            new() { { "activity_type", "deposit" }, { "amount", "100" }, { "coin_type", "0x1::aptos_coin::AptosCoin" } }
        };
        _world.Result = _world.TestVectors["coinActivities"];
    }

    [Then("I should receive deposit and withdrawal activities")]
    public void ThenIShouldReceiveDepositAndWithdrawalActivities()
    {
        var activities = _world.TestVectors["coinActivities"] as List<Dictionary<string, object>>;
        activities.Should().NotBeNull();
    }

    // =========================================================================
    // Processor Status
    // =========================================================================

    [When("I query processor status")]
    public void WhenIQueryProcessorStatus()
    {
        _world.TestVectors["processorStatus"] = new Dictionary<string, object>
        {
            { "processor_name", "token_processor" },
            { "last_success_version", "12345678" },
            { "last_updated", "2024-01-01T00:00:00Z" }
        };
        _world.Result = _world.TestVectors["processorStatus"];
    }

    [Then("I should see which processors are running")]
    public void ThenIShouldSeeWhichProcessorsAreRunning()
    {
        var status = _world.TestVectors["processorStatus"] as Dictionary<string, object>;
        status.Should().ContainKey("processor_name");
    }

    [Then("I should see their last processed versions")]
    public void ThenIShouldSeeTheirLastProcessedVersions()
    {
        var status = _world.TestVectors["processorStatus"] as Dictionary<string, object>;
        status.Should().ContainKey("last_success_version");
    }

    [Given("current ledger version is known")]
    public void GivenCurrentLedgerVersionIsKnown()
    {
        _world.TestVectors["currentLedgerVersion"] = "12345700";
    }

    [When("I compare indexer version to ledger version")]
    public void WhenICompareIndexerVersionToLedgerVersion()
    {
        var status = _world.TestVectors["processorStatus"] as Dictionary<string, object>;
        var indexerVersion = long.Parse(status?["last_success_version"]?.ToString() ?? "12345678");
        var ledgerVersion = long.Parse(_world.TestVectors["currentLedgerVersion"] as string ?? "12345700");
        _world.TestVectors["indexerLag"] = ledgerVersion - indexerVersion;
    }

    [Then("I can determine the lag")]
    public void ThenICanDetermineTheLag()
    {
        var lag = (long)_world.TestVectors["indexerLag"];
        lag.Should().BeGreaterThanOrEqualTo(0);
    }

    // =========================================================================
    // Error Handling
    // =========================================================================

    [Given("the indexer is unavailable")]
    public void GivenTheIndexerIsUnavailable()
    {
        _world.TestVectors["indexerUnavailable"] = true;
    }

    [When("I make a query")]
    public void WhenIMakeAQuery()
    {
        if (_world.TestVectors.TryGetValue("indexerUnavailable", out var iu) && (bool)iu)
        {
            _world.SetError(new Exception("Indexer unavailable"));
        }
    }

    [Then("I should receive a connection error")]
    public void ThenIShouldReceiveAConnectionError()
    {
        _world.Error.Should().NotBeNull();
    }

    [Given("a very slow query")]
    public void GivenAVerySlowQuery()
    {
        _world.TestVectors["slowQuery"] = true;
    }

    [When("it exceeds timeout")]
    public void WhenItExceedsTimeout()
    {
        _world.SetError(new Exception("Query timeout"));
    }

    [Then("I should receive an indexer timeout error")]
    public void ThenIShouldReceiveAnIndexerTimeoutError()
    {
        _world.Error.Should().NotBeNull();
        _world.Error!.Message.Should().Contain("timeout");
    }

    [Given("a query that returns unexpected format")]
    public void GivenAQueryThatReturnsUnexpectedFormat()
    {
        _world.TestVectors["malformedResponse"] = true;
    }

    [When("the response is malformed")]
    public void WhenTheResponseIsMalformed()
    {
        _world.SetError(new Exception("Malformed response"));
    }

    [Then("I should receive a parsing error")]
    public void ThenIShouldReceiveAParsingError()
    {
        _world.Error.Should().NotBeNull();
    }
}

File: tests/python/steps/indexer_steps.py
"""
Step definitions for indexer and GraphQL operations.
"""

from behave import given, when, then

# =============================================================================
# Given Steps - Indexer Setup
# =============================================================================


@given("an indexer client")
def step_given_indexer_client(context):
    # TODO: implement indexer client setup
    context.world.test_vectors["indexer_client"] = True


@given("a custom indexer URL")
def step_given_custom_indexer_url(context):
    context.world.test_vectors["indexer_url"] = "https://custom.indexer.example.com"


@given("an API key for the indexer")
def step_given_api_key(context):
    context.world.test_vectors["api_key"] = "test-api-key"


@given("an unreachable indexer endpoint")
def step_given_unreachable_indexer(context):
    context.world.test_vectors["indexer_url"] = "https://unreachable.invalid"


@given("a GraphQL query string")
def step_given_graphql_query(context):
    context.world.test_vectors["graphql_query"] = "query { ledger_info { chain_id } }"


@given("a GraphQL query with variables")
def step_given_graphql_with_vars(context):
    context.world.test_vectors["graphql_query"] = (
        "query($addr: String!) { account(address: $addr) { sequence_number } }"
    )
    context.world.test_vectors["graphql_vars"] = {"addr": "0x1"}


@given("an invalid GraphQL query")
def step_given_invalid_graphql(context):
    context.world.test_vectors["graphql_query"] = "invalid query {"


@given("a fungible asset query")
def step_given_fa_query(context):
    context.world.test_vectors["fa_query"] = True


@given("an event query result")
def step_given_event_query_result(context):
    context.world.test_vectors["event_result"] = []


@given("an event type")
def step_given_event_type(context):
    context.world.test_vectors["event_type"] = "0x1::coin::DepositEvent"


@given("indexer processor status")
def step_given_processor_status(context):
    context.world.test_vectors["processor_status"] = True


@given("a very complex query")
def step_given_complex_query(context):
    context.world.test_vectors["complex_query"] = True


# =============================================================================
# When Steps - Indexer Operations
# =============================================================================


@when("I create an indexer client for mainnet")
def step_create_indexer_mainnet(context):
    context.world.test_vectors["indexer_network"] = "mainnet"


@when("I create an indexer client for testnet")
def step_create_indexer_testnet(context):
    context.world.test_vectors["indexer_network"] = "testnet"


@when("I create an indexer client with the custom URL")
def step_create_indexer_custom_url(context):
    context.world.test_vectors["indexer_custom"] = True


@when("I create an indexer client with the key")
def step_create_indexer_with_key(context):
    context.world.test_vectors["indexer_with_key"] = True


@when("I execute the query")
def step_execute_query(context):
    # TODO: implement query execution
    context.world.test_vectors["query_executed"] = True


@when("I execute the query with variables")
def step_execute_query_with_vars(context):
    context.world.test_vectors["query_executed"] = True


@when("I query current tokens for the account")
def step_query_current_tokens(context):
    # TODO: implement token query
    context.world.test_vectors["tokens_queried"] = True


@when("I query current tokens")
def step_query_tokens(context):
    context.world.test_vectors["tokens_queried"] = True


@when("I query account transactions")
def step_query_account_txs(context):
    context.world.test_vectors["txs_queried"] = True


@when("I query coin balances from indexer")
def step_query_coin_balances(context):
    context.world.test_vectors["balances_queried"] = True


@when("I query fungible asset balances")
def step_query_fa_balances(context):
    context.world.test_vectors["fa_queried"] = True


@when("I query APT fungible asset balance")
def step_query_apt_balance(context):
    context.world.test_vectors["apt_queried"] = True


@when("I query coin activities")
def step_query_coin_activities(context):
    context.world.test_vectors["activities_queried"] = True


@when("I query events of that type")
def step_query_events_by_type(context):
    context.world.test_vectors["events_queried"] = True


@when("I query events involving that account")
def step_query_events_for_account(context):
    context.world.test_vectors["events_queried"] = True


@when("I query processor status")
def step_query_processor_status(context):
    context.world.test_vectors["processor_queried"] = True


@when("I query it from indexer")
def step_query_from_indexer(context):
    context.world.test_vectors["indexer_queried"] = True


@when("I query the collection")
def step_query_collection(context):
    context.world.test_vectors["collection_queried"] = True


@when("I query the token")
def step_query_token(context):
    context.world.test_vectors["token_queried"] = True


@when("I query tokens in the collection")
def step_query_tokens_in_collection(context):
    context.world.test_vectors["collection_tokens_queried"] = True


@when("I query tokens with limit 10 and offset 0")
def step_query_tokens_paginated(context):
    context.world.test_vectors["tokens_paginated"] = True


@when("I query with limit 25")
def step_query_with_limit(context):
    context.world.test_vectors["query_limit"] = 25


@when("I query with offset 10")
def step_query_with_offset(context):
    context.world.test_vectors["query_offset"] = 10


@when("I query only user transactions")
def step_query_user_txs(context):
    context.world.test_vectors["user_txs_only"] = True


@when("I query its metadata")
def step_query_metadata(context):
    context.world.test_vectors["metadata_queried"] = True


@when("I try to query")
def step_try_query(context):
    try:
        context.world.test_vectors["query_attempted"] = True
    except Exception as e:
        context.world.set_error(e)


# =============================================================================
# Then Steps - Indexer Assertions
# =============================================================================


@then("I should receive the query result")
def step_receive_query_result(context):
    pass  # Generic assertion


@then("I should receive a list of tokens")
def step_receive_token_list(context):
    pass


@then("I should receive a list of transactions")
def step_receive_tx_list(context):
    pass


@then("I should receive a list of balances")
def step_receive_balance_list(context):
    pass


@then("I should receive a list of modules")
def step_receive_module_list(context):
    pass


@then("I should receive all coin types and amounts")
def step_receive_coin_types(context):
    pass


@then("I should receive matching events")
def step_receive_matching_events(context):
    pass


@then("I should receive relevant events")
def step_receive_relevant_events(context):
    pass


@then("I should see deposits and withdrawals")
def step_see_deposits_withdrawals(context):
    pass


@then("I should see the last processed version")
def step_see_last_version(context):
    pass


@then("I can compare with fullnode ledger version")
def step_compare_fullnode_version(context):
    pass


@then("I can determine indexer lag")
def step_determine_indexer_lag(context):
    pass


@then("I should receive a GraphQL error")
def step_receive_graphql_error(context):
    assert context.world.error is not None or context.world.test_vectors.get("error")


@then("I should receive an empty list")
def step_receive_empty_list(context):
    pass


@then("I should only receive user transactions")
def step_only_user_txs(context):
    pass


@then("I should receive at most 25 transactions")
def step_at_most_25_txs(context):
    pass


@then("I should receive at most 5 transactions")
def step_at_most_5_txs(context):
    pass


@then("I should receive at most 10 tokens")
def step_at_most_10_tokens(context):
    pass


@then("I should receive the next page")
def step_receive_next_page(context):
    pass


@then("I should receive tokens belonging to that collection")
def step_receive_collection_tokens(context):
    pass


@then("I should receive collection details")
def step_receive_collection_details(context):
    pass


@then("I should see name")
def step_see_name(context):
    pass


@then("I should see description")
def step_see_description(context):
    pass


@then("I should see uri")
def step_see_uri(context):
    pass


@then("I should see current_supply")
def step_see_current_supply(context):
    pass


@then("I should see token_name")
def step_see_token_name(context):
    pass


@then("I should see collection_name")
def step_see_collection_name(context):
    pass


@then("I should see creator_address")
def step_see_creator_address(context):
    pass


@then("I should see token_uri")
def step_see_token_uri(context):
    pass


@then('I should see name "Aptos Coin"')
def step_see_name_apt(context):
    pass


@then('I should see symbol "APT"')
def step_see_symbol_apt(context):
    pass


@then("I should see decimals 8")
def step_see_decimals_8(context):
    pass


@then("I should see symbol")
def step_see_symbol(context):
    pass


@then("I should see decimals")
def step_see_decimals(context):
    pass


@then("I should see amount")
def step_see_amount(context):
    pass


@then("requests should include the API key header")
def step_requests_include_key(context):
    pass
